File: Samples/iOS/iOS/Moment+Request.m
//
//  Moment+Request.m
//  CoreDataClientSample
//
//  Created by Nathaniel Potter on 11/4/14.
//  Copyright (c) 2014 Nathaniel Potter. All rights reserved.
//

#import "Moment+Request.h"

@implementation Moment (Request)

+ (void)initialize
{
    [self setBasePath:@"https://www.googleapis.com/plus/v1/people/me/moments/vault/"];
}

- (void)saveWithCompletion:(void(^)(Moment *moment, NSError *error))completion
{
    [self POST:@"" progress:nil request:^(NSMutableURLRequest *request) {
        NSError *error;
        NSData *postBody = [NSJSONSerialization dataWithJSONObject:[self moment] options:0 error:&error];
        [request setData:postBody ofContentType:postBodyContentTypeJSON];
    } withCompletion:^(NSArray *results, NSError *error) {
        completion(results.lastObject, error);
    }];
}

- (NSDictionary *)moment
{
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    [dateFormatter setFormatterBehavior:NSDateFormatterBehavior10_4];
    [dateFormatter setDateFormat:@"yyyy-MM-dd'T'HH:mm:ss'Z'"];
    NSString *dateString = [dateFormatter stringFromDate:[NSDate date]];

    NSDictionary *moment = @{@"type":@"http://schema.org/AddAction",
                             @"startDate":dateString,
                             @"object":@{@"id":@"target-id-1",
                                          @"image":@"http://www.google.com/s2/static/images/GoogleyEyes.png",
                                          @"type":@"http://schema.org/AddAction",
                                          @"description":@"The description for the activity",
                                          @"name":@"An example of AddActivity"}
                               };
    NSLog(@"----- %@", moment);
    
    return moment;
                             /*
    var payload = {
        "type":"http:\/\/schemas.google.com\/AddActivity",
        "startDate": "2012-10-31T23:59:59.999Z"
    };
    if (url != undefined){
        payload.target = {
            'url' : 'https://developers.google.com/+/plugins/snippet/examples/thing'
        };
    }else{
        payload.target = {
            "id" : "replacewithuniqueidforaddtarget",
            "image" : "http:\/\/www.google.com\/s2\/static\/images\/GoogleyEyes.png",
            "type" : "http:\/\/schema.org\/CreativeWork",
            "description" : "The description for the activity",
            "name":"An example of AddActivity"
        };
    }
                              */
}

@end

File: Samples/iOS_Swift/CoreDataClientSample/Moment+Request.swift
//
//  Moment+Request.swift
//  CoreDataClientSample
//
//  Created by Nathaniel Potter on 1/31/15.
//  Copyright (c) 2015 Nathaniel Potter. All rights reserved.
//

import Foundation

extension Moment {
    
    struct Initialize {
        static var basePath: String = "https://www.googleapis.com/plus/v1/"
        static var managedObjectUidKey: String = "id"
        static var responseObjectUidKey: String = "id"
    }
    
    class var basePath: String {
        return Initialize.basePath
    }
    
    class var managedObjectUidKey: String {
        return Initialize.managedObjectUidKey
    }
    
    class var responseObjectUidKey: String {
        return Initialize.responseObjectUidKey
    }
    
    func saveMomentWithCompletion(completion: (results: [AnyObject]?, error: NSError?)->()) {
        
        self.POST("people/me/moments/vault", request: { (request: NSMutableURLRequest!) -> Void in
            request.setJSON(self.momentDictionary())
        }) { (results: Array!, error: NSError!) -> Void in
            completion(results: results, error: error)
        }
    }
    
    func deleteMomentWithCompletion(completion: (results: [AnyObject]?, error: NSError?)->()) {
        if let id = self.id {
            let path = "moments/\(id)"
            self.DELETE(path, withCompletion: { (results: Array!, error: NSError!) -> Void in
                completion(results: results, error: error)
            })
        }
    }
    
    /*
    func saveValuesForKeys(keys: [NSString], completion: (results: [AnyObject]?, error: NSError?)->()) {
    
        self.PUT(user.uid, request: { (request: NSMutableURLRequest!) -> Void in
            request.setJSON(self.dictionaryWithValuesForKeys(keys))
            }) { (results: Array!, error: NSError!) -> Void in
                completion(results: results, error: error)
        }
    }
*/
    
    func momentDictionary() -> [NSObject : AnyObject] {
        
        struct DateFormatter {
            static var dateFormatter: NSDateFormatter = {
                let dateFormatter = NSDateFormatter()
                dateFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss'Z'"
                return dateFormatter
                }()
        }
        
        var dateFormatter: NSDateFormatter {
            get { return DateFormatter.dateFormatter }
        }
        
        let dateString = dateFormatter.stringFromDate(NSDate())
        
        let moment: Dictionary = [  "type":"http://schema.org/AddAction",
                                    "startDate":dateString,
                                    "object":[  "id":"target-id-1",
                                                "image":"http://www.fillmurray.com/50/50",
                                                "type":"http://schema.org/AddAction",
                                                "description":"The description for the activity",
                                                "name":"An example of AddActivity"
                                            ]
                                ]
        
        println("\(moment)")
        
        return moment;
    }
}
